src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIRequestSmelter.java

```diff
@@ -218,6 +223,140 @@ private IAIState accelerateFurnaces()
         return getState();
     }
 
+    private List<ItemStack> getActivePossibleFuels()
+    {
+        final List<ItemStack> possibleFuels = getOwnBuilding().getAllowedFuel();
+        if (possibleFuels.isEmpty())
+        {
+            if (worker.getCitizenData() != null)
+            {
+                worker.getCitizenData().triggerInteraction(new StandardInteraction(new TranslationTextComponent(FURNACE_USER_NO_FUEL), ChatPriority.IMPORTANT));
+            }
+            return ImmutableList.of();
+        }
+
+        if(currentRecipeStorage != null)
+        {
+            possibleFuels.removeIf(stack -> ItemStackUtils.compareItemStacksIgnoreStackSize(stack, currentRecipeStorage.getPrimaryOutput()));
+            // There is always only one input.
+            possibleFuels.removeIf(stack -> ItemStackUtils.compareItemStacksIgnoreStackSize(stack, currentRecipeStorage.getCleanedInput().get(0).getItemStack()));
+        }
+        return possibleFuels;
+    }
+
+    private boolean isFuelNeeded()
+    {
+        for (final BlockPos pos : getOwnBuilding().getFurnaces())
+        {
+            if (WorldUtil.isBlockLoaded(world, pos))
+            {
+                final TileEntity entity = world.getTileEntity(pos);
+                final FurnaceTileEntity furnace = (FurnaceTileEntity) entity;
+                if (!furnace.isBurning() && (hasSmeltableInFurnaceAndNoFuel(furnace) || hasNeitherFuelNorSmeltAble(furnace)) && currentRecipeStorage != null && currentRecipeStorage.getIntermediate() == Blocks.FURNACE) 
+                {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Check Fuel levels in the furnace
+     */
+    private IAIState checkFurnaceFuel()
+    {
+        final World world = getOwnBuilding().getColony().getWorld();
+        final List<ItemStack> possibleFuels = getActivePossibleFuels();
+        final int amountOfFuelInBuilding = InventoryUtils.getItemCountInProvider(getOwnBuilding(), item -> FurnaceTileEntity.isFuel(item) && possibleFuels.stream().anyMatch(candidate -> ItemStackUtils.compareItemStacksIgnoreStackSize(candidate, item)));
+        final int amountOfFuelInInv = InventoryUtils.getItemCountInItemHandler(worker.getInventoryCitizen(), item -> FurnaceTileEntity.isFuel(item) && possibleFuels.stream().anyMatch(candidate -> ItemStackUtils.compareItemStacksIgnoreStackSize(candidate, item)));
+
+        if (amountOfFuelInBuilding + amountOfFuelInInv <= 0 && !getOwnBuilding().hasWorkerOpenRequestsOfType(worker.getCitizenData(), TypeToken.of(StackList.class)) && currentRecipeStorage != null && currentRecipeStorage.getIntermediate() == Blocks.FURNACE)
```

Comment:
Not sure it's worth it. But for performance reasons we could always first check the inventory and if it doesn't then we check the building (building count can be fairly expensive). Potentially also just check for hasInInv instead of count here?

Reply:
I agree that it's probably not worth it. Especially with the addition of the isFuelNeeded predicate for the state machine. This cuts down on how often we do the actual inventory checks significantly.

Reply:
how often is this called?

Reply:
new AIEventTarget(AIBlockingEventType.STATE_BLOCKING, this::isFuelNeeded, this::checkFurnaceFuel, TICKS_SECOND),

So, isFuelNeeded is called once a second. Only if it returns true is checkFurnaceFuel called, generally about 3 times for each furnace that needs fueling in the SpaceWars baker, during the walk to fuel it.